File: leet/src/Q_11_20/Q16/Q16.go
package main

import (
	"fmt"
	"math"
	"sort"
)

//执行用时 :4 ms, 在所有 golang 提交中击败了98.14%的用户
//内存消耗 :2.7 MB, 在所有 golang 提交中击败了61.29%的用户
func threeSumClosest(nums []int, target int) int {
	sort.Ints(nums)
	ans := nums[0] + nums[1] + nums[2]
	for i := 0; i < len(nums); i++ {
		start := i + 1
		ends := len(nums) - 1
		for start < ends {
			sum := nums[i] + nums[start] + nums[ends]
			tem := int(math.Abs(float64(sum) - float64(target)))
			current := int(math.Abs(float64(ans) - float64(target)))
			if tem < current {
				ans = sum
			}
			if sum > target {
				ends--
			} else if sum < target {
				start++
			} else {
				return ans
			}
		}
	}
	return ans
}
func main() {
	c := []int{-1, 1, 2, -4}
	fmt.Println(threeSumClosest(c, 1))
}

File: leet/src/Q_11_20/Q16/Q16.java
package Q_11_20.Q16;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author bockey
 * <p>
 * <p>
 * 执行用时 :6 ms, 在所有 java 提交中击败了89.47%的用户
 * 内存消耗 :36.5 MB, 在所有 java 提交中击败了84.18%的用户
 */
public class Q16 {
    public int threeSumClosest(int[] nums, int target) {
        int ans = nums[0] + nums[1] + nums[2];
        Arrays.sort(nums);
        for (int i = 0; i < nums.length; i++) {
            int start = i + 1;
            int ends = nums.length - 1;
            while (start < ends) {
                int sum = nums[i] + nums[start] + nums[ends];
                int tem = (Math.abs(sum - target));
                int current = (Math.abs(target - ans));
                if (tem < current) ans = sum;

                if (sum > target) ends--;
                else if (sum < target) start++;
                else return ans;
            }
        }
        return ans;
    }

    public static void main(String[] args) {
//        int[] ss = {1, 2, 3, 4};[-1，2，1，-4]
        int[] ss = {0, 1, 2};
//        int[] ss = {-1, 1, 2, -4};
        System.out.println(new Q16().threeSumClosest(ss, 3));
    }
}

